packages/plugin-react/src/fast-refresh.ts

```diff
@@ -0,0 +1,109 @@
+import type { types as t } from '@babel/core'
+import fs from 'fs'
+
+export const runtimePublicPath = '/@react-refresh'
+
+const runtimeFilePath = require.resolve(
+  'react-refresh/cjs/react-refresh-runtime.development.js'
+)
+
+export const runtimeCode = `
+const exports = {}
+${fs.readFileSync(runtimeFilePath, 'utf-8')}
+function debounce(fn, delay) {
+  let handle
+  return () => {
+    clearTimeout(handle)
+    handle = setTimeout(fn, delay)
+  }
+}
+exports.performReactRefresh = debounce(exports.performReactRefresh, 16)
+export default exports
+`
+
+export const preambleCode = `
+import RefreshRuntime from "__BASE__${runtimePublicPath.slice(1)}"
+RefreshRuntime.injectIntoGlobalHook(window)
+window.$RefreshReg$ = () => {}
+window.$RefreshSig$ = () => (type) => type
+window.__vite_plugin_react_preamble_installed__ = true
+`
+
+const header = `
+import RefreshRuntime from "${runtimePublicPath}";
+
+let prevRefreshReg;
+let prevRefreshSig;
+
+if (!window.__vite_plugin_react_preamble_installed__) {
+  throw new Error(
+    "@vitejs/plugin-react-refresh can't detect preamble. Something is wrong. " +
+    "See https://github.com/vitejs/vite-plugin-react/pull/11#discussion_r430879201"
+  );
+}
+
+if (import.meta.hot) {
+  prevRefreshReg = window.$RefreshReg$;
+  prevRefreshSig = window.$RefreshSig$;
+  window.$RefreshReg$ = (type, id) => {
+    RefreshRuntime.register(type, __SOURCE__ + " " + id)
+  };
+  window.$RefreshSig$ = RefreshRuntime.createSignatureFunctionForTransform;
+}`.replace(/[\n]+/gm, '')
+
+const footer = `
+if (import.meta.hot) {
+  window.$RefreshReg$ = prevRefreshReg;
+  window.$RefreshSig$ = prevRefreshSig;
+
+  __ACCEPT__
+  if (!window.__vite_plugin_react_timeout) {
+    window.__vite_plugin_react_timeout = setTimeout(() => {
+      window.__vite_plugin_react_timeout = 0;
+      RefreshRuntime.performReactRefresh();
+    }, 30);
+  }
+}`
+
+export function addRefreshWrapper(
+  code: string,
+  id: string,
+  accept: boolean
+): string {
+  return (
+    header.replace('__SOURCE__', JSON.stringify(id)) +
+    code +
+    footer.replace('__ACCEPT__', accept ? 'import.meta.hot.accept();' : '')
+  )
+}
+
+export function isRefreshBoundary(ast: t.File): boolean {
+  // Every export must be a React component.
+  return ast.program.body.every((node) => {
+    if (node.type !== 'ExportNamedDeclaration') {
```

Comment:
It looks like we ignore direct exports, no matter is anonymous or not. Should we need to follow the strategy mentioned in facebook/react#21181.

Reply:
I'm not sure. Maybe this can be addressed in a follow-up PR.